components/Wizard/formComponents/Packages.js

```diff
@@ -0,0 +1,537 @@
+import useFieldApi from "@data-driven-forms/react-form-renderer/use-field-api";
+import PropTypes from "prop-types";
+import {
+  Bullseye,
+  Button,
+  DualListSelector,
+  DualListSelectorPane,
+  DualListSelectorList,
+  DualListSelectorListItem,
+  DualListSelectorControlsWrapper,
+  DualListSelectorControl,
+  SearchInput,
+  Spinner,
+  TextContent,
+} from "@patternfly/react-core";
+import { AngleDoubleLeftIcon, AngleLeftIcon, AngleDoubleRightIcon, AngleRightIcon } from "@patternfly/react-icons";
+import { connect } from "react-redux";
+import api from "../../../core/api";
+import * as composer from "../../../core/composer";
+import { updateBlueprintComponents } from "../../../core/actions/blueprints";
+
+// the fields isHidden and isSelected should not be included in the package list sent for image creation
+const removePackagesDisplayFields = (packages) =>
+  packages.map((pack) => ({
+    name: pack.name,
+    summary: pack.summary,
+  }));
+
+const Packages = ({ defaultArch, ...props }) => {
+  const { change, getState } = useFormApi();
+  const { input } = useFieldApi(props);
+  const [packagesSearchName, setPackagesSearchName] = useState(undefined);
+  const [filterAvailable, setFilterAvailable] = useState(undefined);
+  const [filterChosen, setFilterChosen] = useState(undefined);
+  const [packagesAvailable, setPackagesAvailable] = useState([]);
+  const [packagesAvailableFound, setPackagesAvailableFound] = useState(true);
+  const [packagesAvailableLoading, setPackagesAvailableLoading] = useState(false);
+  const [packagesChosen, setPackagesChosen] = useState([]);
+  const [packagesChosenFound, setPackagesChosenFound] = useState(true);
+  const [packagesChosenLoading, setPackagesChosenLoading] = useState(false);
+  const [focus, setFocus] = useState("");
+  const [totalPackagesAvailable, setTotalPackagesAvailable] = useState(undefined);
+  const [packagesAvailablePage, setPackagesAvailablePage] = useState(1);
+  const [packagesAvailablePerPage, setPackagesAvailablePerPage] = useState(100);
+  const [additionalPackagesAvailableLoading, setAdditionalPackagesAvailableLoading] = useState(false);
+
+  // this effect only triggers on mount
+  useEffect(() => {
+    const fetchPackagesChosen = async () => {
+      setPackagesChosenLoading(true);
+      const result = await composer.depsolveBlueprint(props.blueprintName);
+      const blueprint = result.blueprints[0].blueprint;
+      const blueprintPackages = blueprint.packages.concat(blueprint.modules);
+      updateBlueprintComponents(blueprintPackages);
+      const packageNames = blueprintPackages.map((pkg) => pkg.name);
+      if (packageNames.length !== 0) {
+        const packageInfo = await composer.getComponentInfo(packageNames);
+        const selectedPackages = packageInfo.map((pkg) => ({ name: pkg.name, summary: pkg.summary }));
+        setPackagesChosenSorted(selectedPackages);
+        change(input.name, removePackagesDisplayFields(selectedPackages));
```

Comment:
no need to change the field parameter like this. The packages with name and summary isn't used elsewhere. Instead, you could set the packages form field to the list of package names. That would enable you, on the review page, to load the packages names from the form state and not the blueprint.

Reply:
Good point! This does add some additional complexity to the review page, though, since the package list is fetched using an async API call. If a user clicks 'next' on the Packages step before the packages are loaded, there is a problem... when the API call completes the form state will be set correctly but the review step won't re-render to reflect this. My solution to this is to use a <FormSpy> to keep track of changes to the form state and re-render the review page when they occur.